File: ruby/0057-insert-interval.rb
def insert(intervals, new_interval)
    result = []
    (0..intervals.size()-1).each do |i|
        if new_interval[1] < intervals[i][0]
            result.append(new_interval)
            result += intervals[i..]
            return result
        elsif new_interval[0] > intervals[i][1]
            result.append(intervals[i])
        else
            new_interval = [[new_interval[0],intervals[i][0]].min, [new_interval[1],intervals[i][1]].max]
        end
    end
    result.append(new_interval)
    return result
end

File: typescript/0057-insert-interval.ts
function insert(intervals: number[][], newInterval: number[]): number[][] {
    const { beforeIndex, before } = getBefore(intervals, newInterval);
    const afterIndex = mergeIntervals(intervals, newInterval, beforeIndex);
    const after = intervals.slice(afterIndex);

    return [...before, newInterval, ...after]; 
};

const getBefore = (intervals, newInterval, index = 0, before = []) => {
    const hasGap = ([prevStart, prevEnd], [currStart, currEnd]) => prevEnd < currStart;
    
    while (index < intervals.length && hasGap(intervals[index], newInterval)) {
        const current = intervals[index];
        before.push(current);
        index++;
    }
    return { beforeIndex: index, before };
};

const mergeIntervals = (intervals, newInterval, index) => {
    const hasOverlap = ([prevStart, prevEnd], [currStart, currEnd]) => currStart <= prevEnd;

    while (index < intervals.length && hasOverlap(newInterval, intervals[index])) {
        const current = intervals[index];
        newInterval[0] = Math.min(newInterval[0], current[0]);
        newInterval[1] = Math.max(newInterval[1], current[1]);
        index++;
    }
    return index;
};